File: Haskell/PBM.hs
-- PBM.hs
-- Carga de archivos PBM P4 y acceso a pixeles individuales.
-- Responsabilidad de este modulo: transformar bytes crudos del archivo
-- en una estructura (Imagen) y ofrecer una funcion pura para saber si
-- un pixel (x, y) es negro o blanco.

module PBM
  ( Imagen(..)
  , cargarPBM
  , bytesPorFila
  , esNegro
  ) where

import qualified Data.ByteString as BS
import Data.ByteString (ByteString)
import Data.Bits (testBit)
import Data.Word (Word8)
import Data.Char (chr, isSpace, isDigit)

-- | Una imagen PBM P4 ya interpretada: dimensiones + datos binarios crudos
-- (los datos siguen empacados en bits, tal como vienen en el archivo).
data Imagen = Imagen
  { ancho   :: Int
  , alto    :: Int
  , pixeles :: ByteString
  } deriving (Show)

-- | Lee el archivo del disco y construye la Imagen.
cargarPBM :: FilePath -> IO Imagen
cargarPBM ruta = do
  contenido <- BS.readFile ruta
  return (parseHeader contenido)

-- | Interpreta el header "P4\n[comentarios]\nANCHO ALTO\n<datos binarios>"
-- y separa las dimensiones de los datos de pixeles.
parseHeader :: ByteString -> Imagen
parseHeader bs0 =
  let bs1 = requerirMagic bs0
      bs2 = saltarEspaciosYComentarios bs1
      (w, bs3) = leerEntero bs2
      bs4 = saltarEspaciosYComentarios bs3
      (h, bs5) = leerEntero bs4
      -- Tras el ultimo numero del header hay exactamente UN caracter
      -- de espacio en blanco antes de que empiecen los datos binarios.
      datos = BS.drop 1 bs5
  in Imagen { ancho = w, alto = h, pixeles = datos }

-- | Verifica y descarta la marca magica "P4".
requerirMagic :: ByteString -> ByteString
requerirMagic bs =
  case BS.stripPrefix (byteStringDeString "P4") bs of
    Just resto -> resto
    Nothing    -> error "Archivo no es un PBM P4 valido (falta la marca P4)"

-- | Salta espacios en blanco y comentarios ("#" hasta fin de linea),
-- repitiendo hasta encontrar el inicio de un token real.
saltarEspaciosYComentarios :: ByteString -> ByteString
saltarEspaciosYComentarios bs
  | BS.null bs = bs
  | esEspacioByte (BS.head bs) = saltarEspaciosYComentarios (BS.drop 1 bs)
  | BS.head bs == comentarioByte =
      saltarEspaciosYComentarios (BS.dropWhile (/= saltoLineaByte) bs)
  | otherwise = bs

-- | Lee un entero en ASCII (secuencia de digitos) desde el inicio del ByteString.
leerEntero :: ByteString -> (Int, ByteString)
leerEntero bs =
  let (digitos, resto) = BS.span esDigitoByte bs
  in (read (map (chr . fromIntegral) (BS.unpack digitos)), resto)

esEspacioByte :: Word8 -> Bool
esEspacioByte w = isSpace (chr (fromIntegral w))

esDigitoByte :: Word8 -> Bool
esDigitoByte w = isDigit (chr (fromIntegral w))

comentarioByte :: Word8
comentarioByte = fromIntegral (fromEnum '#')

saltoLineaByte :: Word8
saltoLineaByte = fromIntegral (fromEnum '\n')

byteStringDeString :: String -> ByteString
byteStringDeString = BS.pack . map (fromIntegral . fromEnum)

-- | Numero de bytes necesarios para representar una fila de "w" pixeles,
-- ya que PBM P4 empaqueta 8 pixeles (bits) por byte.
bytesPorFila :: Int -> Int
bytesPorFila w = (w + 7) `div` 8

-- | True si el pixel (x, y) de la imagen es negro (bit = 1).
-- El bit mas significativo del byte corresponde al pixel mas a la
-- izquierda de ese grupo de 8.
esNegro :: Imagen -> Int -> Int -> Bool
esNegro img x y =
  let bpf         = bytesPorFila (ancho img)
      indiceByte  = y * bpf + x `div` 8
      byte        = BS.index (pixeles img) indiceByte
      posicionBit = 7 - (x `mod` 8)
  in testBit byte posicionBit

File: Prolog/pbm.pl
% pbm.pl
% Lectura de archivos PBM P4 y acceso a pixeles individuales.
% Responsabilidad de este modulo: leer el archivo como bytes crudos,
% separar el header (ancho, alto) de los datos binarios, y ofrecer una
% relacion que dice si un pixel (X, Y) es negro o blanco.
%
% Los datos se guardan como un string de SWI-Prolog leido con
% encoding(octet): cada "caracter" del string es exactamente un byte
% (codigo 0-255), lo que da acceso O(1) via string_code/3 sin pagar el
% costo de una lista enlazada para una imagen de miles de bytes.

:- module(pbm, [leer_pbm/3, bytes_por_fila/2, pixel_negro/4, byte_de/3]).

% leer_pbm(+Ruta, -Info, -Datos)
% Info es el compound imagen(Ancho, Alto). Datos es el string de bytes
% binarios (ya sin el header), listo para acceso aleatorio.
leer_pbm(Ruta, imagen(Ancho, Alto), Datos) :-
    read_file_to_string(Ruta, Contenido, [encoding(octet)]),
    string_length(Contenido, Total),
    LargoMuestra is min(200, Total),
    sub_string(Contenido, 0, LargoMuestra, _, Muestra),
    string_codes(Muestra, CodigosHeader),
    parse_header(CodigosHeader, Ancho, Alto, Consumido),
    LargoDatos is Total - Consumido,
    sub_string(Contenido, Consumido, LargoDatos, 0, Datos).

% parse_header(+Codigos, -Ancho, -Alto, -Consumido)
% Interpreta "P4" [comentarios/espacios] Ancho [comentarios/espacios] Alto
% seguido de exactamente un caracter de espacio antes de los datos.
% Consumido es el numero de bytes que ocupa el header completo.
parse_header(Codigos, Ancho, Alto, Consumido) :-
    Codigos = [0'P, 0'4 | Resto0],
    saltar_espacios_comentarios(Resto0, Resto1),
    leer_entero(Resto1, Ancho, Resto2),
    saltar_espacios_comentarios(Resto2, Resto3),
    leer_entero(Resto3, Alto, Resto4),
    Resto4 = [_UnEspacio | RestoDatos],
    length(Codigos, LargoTotal),
    length(RestoDatos, LargoRestante),
    Consumido is LargoTotal - LargoRestante.

% Salta espacios en blanco y comentarios ("#" hasta fin de linea),
% repitiendo hasta encontrar el inicio de un token real.
saltar_espacios_comentarios([C | Cs], Resto) :-
    code_type(C, space), !,
    saltar_espacios_comentarios(Cs, Resto).
saltar_espacios_comentarios([0'# | Cs], Resto) :-
    !,
    saltar_linea(Cs, Cs2),
    saltar_espacios_comentarios(Cs2, Resto).
saltar_espacios_comentarios(Codigos, Codigos).

saltar_linea([0'\n | Cs], Cs) :- !.
saltar_linea([_ | Cs], Resto) :- !, saltar_linea(Cs, Resto).
saltar_linea([], []).

% Lee una secuencia de digitos ASCII y la convierte a entero.
leer_entero(Codigos, Valor, Resto) :-
    leer_digitos(Codigos, Digitos, Resto),
    number_codes(Valor, Digitos).

leer_digitos([C | Cs], [C | Ds], Resto) :-
    code_type(C, digit), !,
    leer_digitos(Cs, Ds, Resto).
leer_digitos(Codigos, [], Codigos).

% bytes_por_fila(+Ancho, -BytesPorFila)
% Cada byte empaqueta 8 pixeles (bits); se redondea hacia arriba.
bytes_por_fila(Ancho, BytesPorFila) :-
    BytesPorFila is (Ancho + 7) // 8.

% byte_de(+Datos, +Indice0, -Byte)
% Byte en la posicion Indice0 (base 0) del string de datos.
byte_de(Datos, Indice0, Byte) :-
    Indice1 is Indice0 + 1,
    string_code(Indice1, Datos, Byte).

% pixel_negro(+X, +Y, +Info, +Datos)
% Verdadero si el pixel (X, Y) es negro (bit = 1). El bit mas
% significativo de un byte corresponde al pixel mas a la izquierda
% del grupo de 8.
pixel_negro(X, Y, imagen(Ancho, _Alto), Datos) :-
    bytes_por_fila(Ancho, BytesPorFila),
    IndiceByte is Y * BytesPorFila + X // 8,
    byte_de(Datos, IndiceByte, Byte),
    PosicionBit is 7 - (X mod 8),
    Bit is (Byte >> PosicionBit) /\ 1,
    Bit =:= 1.
